crates/uv-publish/src/lib.rs

```diff
@@ -339,26 +353,123 @@ pub async fn upload(
             PublishError::PublishSend(file.to_path_buf(), registry.clone(), err.into())
         })?;
 
-        return handle_response(registry, response)
+        return match handle_response(registry, response).await {
+            Ok(()) => {
+                // Upload successful; for PyPI this can also mean a hash match in a raced upload
+                // (but it doesn't tell us), for other registries it should mean a fresh upload.
+                Ok(true)
+            }
+            Err(err) => {
+                if matches!(
+                    err,
+                    PublishSendError::Status(..) | PublishSendError::StatusNoBody(..)
+                ) && skip_existing(index_client, index_capabilities, file, filename).await?
+                {
+                    // There was a raced upload of the same file, so even though our upload failed,
+                    // the right file now exists in the registry.
+                    return Ok(false);
+                }
+                Err(PublishError::PublishSend(
+                    file.to_path_buf(),
+                    registry.clone(),
+                    err,
+                ))
+            }
+        };
+    }
+}
+
+/// Check whether we should skip the upload of a file because it already exists on the index.
+pub async fn skip_existing(
+    index_client: &mut Option<(IndexUrl, RegistryClient)>,
+    index_capabilities: &IndexCapabilities,
+    file: &Path,
+    filename: &DistFilename,
+) -> Result<bool, PublishError> {
+    let Some((index_url, registry_client)) = index_client.as_mut() else {
+        return Ok(false);
+    };
+
+    // Avoid using the PyPI 10min default cache.
+    registry_client.refresh(Refresh::from_args(None, vec![filename.name().clone()]));
+
+    debug!("Checking for {filename} in the registry");
+    let response = registry_client
+        .simple(filename.name(), Some(index_url), index_capabilities)
```

Comment:
Do we need the index url here or will it do the right thing without it?

Reply:
It depends. Was the client created with _only_ this index as the set of locations? It seems right to pass it in though.